lora/redis/streams.go

```diff
@@ -0,0 +1,218 @@
+package redis
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/go-redis/redis"
+	"github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/lora"
+)
+
+const (
+	protocol = "lora"
+
+	group  = "mainflux.lora"
+	stream = "mainflux.things"
+
+	thingPrefix = "thing."
+	thingCreate = thingPrefix + "create"
+	thingUpdate = thingPrefix + "update"
+	thingRemove = thingPrefix + "remove"
+
+	channelPrefix = "channel."
+	channelCreate = channelPrefix + "create"
+	channelUpdate = channelPrefix + "update"
+	channelRemove = channelPrefix + "remove"
+)
+
+// EventStore represents event source for things and channels provisioning.
+type EventStore interface {
+	// Subscribes to geven subject and receives events.
+	Subscribe(string)
+}
+
+type thingLoraMetadata struct {
+	Type   string `json:"type"`
+	DevEUI string `json:"devEUI"`
+}
+
+type channelLoraMetadata struct {
+	Type  string `json:"type"`
+	AppID string `json:"appID"`
+}
+
+type eventStore struct {
+	svc      lora.Service
+	client   *redis.Client
+	consumer string
+	logger   logger.Logger
+}
+
+// NewEventStore returns new event store instance.
+func NewEventStore(svc lora.Service, client *redis.Client, consumer string, log logger.Logger) EventStore {
+	return eventStore{
+		svc:      svc,
+		client:   client,
+		consumer: consumer,
+		logger:   log,
+	}
+}
+
+func (es eventStore) Subscribe(subject string) {
+	es.client.XGroupCreate(stream, group, "$").Err()
+	for {
+		streams, err := es.client.XReadGroup(&redis.XReadGroupArgs{
+			Group:    group,
+			Consumer: es.consumer,
+			Streams:  []string{stream, ">"},
+			Count:    100,
+		}).Result()
+		if err != nil || len(streams) == 0 {
+			continue
+		}
+
+		for _, msg := range streams[0].Messages {
+			event := msg.Values
+
+			var err error
+			switch event["operation"] {
+			case thingCreate:
+				cte := decodeCreateThing(event)
+				err = es.handleCreateThing(cte)
+			case thingUpdate:
+				ute := decodeUpdateThing(event)
+				err = es.handleUpdateThing(ute)
+			case thingRemove:
+				rte := decodeRemoveThing(event)
+				err = es.handleRemoveThing(rte)
+			case channelCreate:
+				cce := decodeCreateChannel(event)
+				err = es.handleCreateChannel(cce)
+			case channelUpdate:
+				uce := decodeUpdateChannel(event)
+				err = es.handleUpdateChannel(uce)
+			case channelRemove:
+				rce := decodeRemoveChannel(event)
+				err = es.handleRemoveChannel(rce)
```

Comment:
Add `default:` case where you will put your custom error, something like: `ErrUnprocessableEvent`.